src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -971,6 +966,88 @@ public override void DidReceiveChallenge (NSUrlSession session, NSUrlSessionTask
+			{
+				var certificateValidationCallback = sessionHandler.ServerCertificateCustomValidationCallback;
+				if (certificateValidationCallback is null)
+					throw new InvalidOperationException ($"{nameof(NSUrlSessionHandler.ServerCertificateCustomValidationCallback)} cannot be null");
+
+				X509Certificate2[] certificates = ConvertCertificates (secTrust);
+				X509Certificate2? certificate = certificates.Length > 0 ? certificates [0] : null;
+				X509Chain chain = CreateChain (certificates);
+				SslPolicyErrors sslPolicyErrors = EvaluateSslPolicyErrors (certificate, chain, secTrust);
+
+				return certificateValidationCallback (request, certificate, chain, sslPolicyErrors);
+			}
+
+			X509Certificate2[] ConvertCertificates (SecTrust secTrust)
+			{
+				var certificates = new X509Certificate2 [secTrust.Count];
+
+				if (IsSecTrustGetCertificateChainSupported) {
+					var originalChain = secTrust.GetCertificateChain ();
+					for (int i = 0; i < originalChain.Length; i++)
+						certificates [i] = originalChain [i].ToX509Certificate2 ();
+				} else {
+					for (int i = 0; i < secTrust.Count; i++)
+						certificates [i] = secTrust [i].ToX509Certificate2 ();
+				}
+
+				return certificates;
+			}
+
+			static bool? isSecTrustGetCertificateChainSupported = null;
+			static bool IsSecTrustGetCertificateChainSupported {
+				get {
+					if (!isSecTrustGetCertificateChainSupported.HasValue) {
+#if MONOMAC
+						isSecTrustGetCertificateChainSupported = ObjCRuntime.SystemVersion.CheckmacOS (12, 0);
+#elif WATCH
+						isSecTrustGetCertificateChainSupported = ObjCRuntime.SystemVersion.CheckWatchOS (8, 0);
+#elif IOS || TVOS || MACCATALYST
+						isSecTrustGetCertificateChainSupported = ObjCRuntime.SystemVersion.CheckiOS (15, 0);
+#else
+						#error Unknown platform
+#endif
+					}
+
+					return isSecTrustGetCertificateChainSupported.Value;
+				}
+			}
+
+			X509Chain CreateChain (X509Certificate2[] certificates)
+			{
+				// the chain initialization is based on dotnet/runtime implementation in System.Net.Security.SecureChannel
+				var chain = new X509Chain ();
+				chain.ChainPolicy.RevocationMode = X509RevocationMode.Online;
+				chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+				chain.ChainPolicy.ExtraStore.AddRange (certificates);
+				return chain;
+			}
+
+			SslPolicyErrors EvaluateSslPolicyErrors (X509Certificate2? certificate, X509Chain chain, SecTrust secTrust)
+			{
+				var sslPolicyErrors = SslPolicyErrors.None;
+				
+				try {
+					if (certificate is null) {
+						sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNotAvailable;
+					} else if (!chain.Build (certificate)) {
+						sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+					}
+				} catch {
+					sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+				}
+
+				if (!secTrust.Evaluate (out _)) {
+					sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+				}
```

Comment:
Yes, the try-catch block is there because of the `chain.Build(...)` which can throw an exception when the certificate is invalid (see https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509chain.build?view=net-6.0). The `Evaluate` method isn't supposed to throw (according to https://docs.microsoft.com/en-us/dotnet/api/security.sectrust.evaluate?view=xamarin-ios-sdk-12).

It could be in the try-catch block (if chain.Build throws, there's nothing we really gain by invoking Evaluate, the `sslPolicyErrors` will include `RemoteCertificateChainErrors` anyway), but it's not linked to the chain building code, so I just kept it separately. It would be probably cleaner to extract the chain building logic into a separate method, but the file is already quite long and I don't think it would be that much of an improvement after all.

Let me know if you think this piece of code should be refactored a bit.

Reply:
> Let me know if you think this piece of code should be refactored a bit.

I found it slightly odd to see a catch-all exception handler in the method (which made it seem that the method shouldn't throw any exceptions), and then a block of code outside of it which could go inside without any change in behavior (except if that code threw an exception). In other words: I saw a weird pattern and just wanted to make sure it was intentional and not accidental :)

So no need to change anything, the current code is fine.

Reply:
You're right that the catch-all is suboptimal. I think this should be changed to catch the specific exception that `Build` can throw.